backend/storage-service/src/jvmMain/kotlin/file/services/SynchronizationService.kt

```diff
@@ -0,0 +1,308 @@
+                    )
+                }
+
+                val id = UUID.randomUUID().toString()
+                val device = chooseFolderDevice(session)
+                affectedDevices.add(device)
+
+                Mounts.mount.call(
+                    MountRequest(
+                        listOf(MountFolder(id, folder.path))
+                    ),
+                    authenticatedClient
+                ).orRethrowAs {
+                    throw RPCException.fromStatusCode(HttpStatusCode.InternalServerError, "Failed to prepare folder for synchronization")
+                }
+
+                ids.add(id)
+                devices.add(device)
+                syncTypes.add(syncType)
+            }
+
+            session.sendPreparedStatement(
+                {
+                    setParameter("ids", ids)
+                    setParameter("devices", devices.map { it.id })
+                    setParameter("paths", request.items.map { folder -> folder.path })
+                    setParameter("users", ids.map { actor.username })
+                    setParameter("access", syncTypes.map { it.name })
+                },
+                """
+                    insert into storage.synchronized_folders(
+                        id, 
+                        device_id, 
+                        path,
+                        user_id,
+                        access_type
+                    ) values (
+                        unnest(:ids::text[]),
+                        unnest(:devices::text[]),
+                        unnest(:paths::text[]),
+                        unnest(:users::text[]),
+                        unnest(:access::text[])
+                    )
+            """
+            ).rowsAffected
+        }
+
+        affectedDevices.forEach { device ->
+            val mounter = Mounts.ready.call(Unit, authenticatedClient)
+            if (mounter.statusCode != HttpStatusCode.OK || !mounter.orThrow().ready || !syncthing.isReady(device)) {
+                throw RPCException(
+                    "The synchronization feature is offline at the moment. Your folders will be synchronized when it returns online.",
+                    HttpStatusCode.InternalServerError
+                )
+            }
+        }
+
+        if (affectedRows > 0) {
+            syncthing.writeConfig(affectedDevices.toList())
+        }
+    }
+
+    suspend fun removeFolder(actor: Actor, request: SynchronizationRemoveFolderRequest) {
+        val affectedRows = db.withSession { session ->
+            session.sendPreparedStatement(
+                {
+                    setParameter("ids", request.items.map { it.id })
+                    setParameter("user", actor.username)
+                },
+                """
+                    delete from storage.synchronized_folders
+                    where id in (select unnest(:ids::text[])) and user_id = :user
+                """
+            ).rowsAffected
+        }
+
+        Mounts.unmount.call(
```

Comment:
Exception is thrown outside of the transaction, this will probably give some very confusing results. Also what happens if the mounter is down in this case?

Reply:
I've updated it to do the write to Syncthing before attempting to unmount. I think if the mounter is down, it's important that Syncthing at least is in the right state.

Reply:
I've changed this again, by just moving the unmount call inside the transaction.